index/scorch/merge.go

```diff
@@ -132,37 +132,54 @@ func (s *Scorch) planMergeAtSnapshot(ourSnapshot *IndexSnapshot) error {
 			if segSnapshot, ok := planSegment.(*SegmentSnapshot); ok {
 				oldMap[segSnapshot.id] = segSnapshot
 				if zapSeg, ok := segSnapshot.segment.(*zap.Segment); ok {
-					segmentsToMerge = append(segmentsToMerge, zapSeg)
+					if segSnapshot.LiveSize() == 0 {
+						oldMap[segSnapshot.id] = nil
+					} else {
+						segmentsToMerge = append(segmentsToMerge, zapSeg)
+					}
 					docsToDrop = append(docsToDrop, segSnapshot.deleted)
 				}
 			}
 		}
 
-		filename := zapFileName(newSegmentID)
-		s.markIneligibleForRemoval(filename)
-		path := s.path + string(os.PathSeparator) + filename
-		newDocNums, err := zap.Merge(segmentsToMerge, docsToDrop, path, DefaultChunkFactor)
-		if err != nil {
-			s.unmarkIneligibleForRemoval(filename)
-			return fmt.Errorf("merging failed: %v", err)
-		}
-		segment, err := zap.Open(path)
-		if err != nil {
-			s.unmarkIneligibleForRemoval(filename)
-			return err
-		}
-		sm := &segmentMerge{
-			id:            newSegmentID,
-			old:           oldMap,
-			oldNewDocNums: make(map[uint64][]uint64),
-			new:           segment,
-			notify:        make(notificationChan),
-		}
-		notifications = append(notifications, sm.notify)
-		for i, segNewDocNums := range newDocNums {
-			sm.oldNewDocNums[task.Segments[i].Id()] = segNewDocNums
+		var sm *segmentMerge
+		if len(segmentsToMerge) > 0 {
+			filename := zapFileName(newSegmentID)
+			s.markIneligibleForRemoval(filename)
+			path := s.path + string(os.PathSeparator) + filename
+			newDocNums, err := zap.Merge(segmentsToMerge, docsToDrop, path, 1024)
+			if err != nil {
+				s.unmarkIneligibleForRemoval(filename)
+				return fmt.Errorf("merging failed: %v", err)
+			}
+			segment, err := zap.Open(path)
+			if err != nil {
+				s.unmarkIneligibleForRemoval(filename)
+				return err
+			}
+			sm = &segmentMerge{
+				id:            newSegmentID,
+				old:           oldMap,
+				oldNewDocNums: make(map[uint64][]uint64),
+				new:           segment,
+				notify:        make(notificationChan),
+			}
+
+			for i, segNewDocNums := range newDocNums {
+				sm.oldNewDocNums[task.Segments[i].Id()] = segNewDocNums
+			}
+
+		} else if len(task.Segments) > 0 {
```

Comment:
this else if case seems to append a very similar structure, seems like the code could be rearranged to do this a simpler way.